### PR TITLE
Unix library

### PR DESCRIPTION
**Pull request by hauptmech**
Illustrates building the unix port as a library with minimal changes for the discussion in #2075

### CODE REVIEW

**Comment by stinos** [comment]
Let's check with pfalcon and dpgeorge, but I think one objection to this patch is going to be that you removed compile_only without explaining why, and did so in the same commit where 2 other basically unrelated changes were also made (adding emit_opt as argument and using printf): that makes things hard to follow (each commit should do one thing if possible) and raises more questions than it answers. I don't recall all reasons exactly, but there definitely were a couple of compelling ones for not using printf but mp_printf instead. And a nitpick: seems like you just copied includes around, so now both main.c and unixport.c probably have includes they don't use which goes against the general 'only include what you need' rule.

**Comment by hauptmech** [comment]
Great points I agree with. The patch is meant only to support the discussion.

Regarding compile_only, what is its output? Error messages only or is there a magic saving of emitted code to file that I'm not seeing?

**Comment by hauptmech** [comment]
> I don't recall all reasons exactly, but there definitely were a couple of compelling ones for not using printf but mp_printf instead.

`git log -L 545,545:main.c`

```diff
commit 3a6b3d230c60c3acfb212cec2bd80305232af42c
Author: Paul Sokolovsky <pfalcon@users.sourceforge.net>
Date:   Fri Nov 13 15:28:57 2015 +0200

    main.c: Switch stderr printing from ANSI C to native POSIX.

diff --git a/unix/main.c b/unix/main.c
--- a/unix/main.c
+++ b/unix/main.c
@@ -518,2 +520,2 @@
-                fprintf(stderr, "%s: can't open file '%s': [Errno %d] ", argv[0], argv[a], errno);
+                mp_printf(&mp_stderr_print, "%s: can't open file '%s': [Errno %d] ", argv[0], argv[a
], errno);
                 perror("");
```

Which points to #1500 ...

But shouldn't @pfalcon have used the API here?

```
    DEBUG_printf("%s: can't open file '%s': [Errno %d] ", argv[0], argv[a], errno);
```

**Comment by dpgeorge** [comment]
The unix/ port already does too much stuff (it's also the basis for windows/) so adding this just makes it more complicated.  I'd rather have a new top-level directory to build such a library.

**Comment by hauptmech** [comment]
Pick a directory name, drop a couple notes on the general vision here or in a readme, and link to your hack script branch for renaming/context insertion in your personal repo. (I scanned your repo but did not see it, maybe you did not push it?)

**Comment by pfalcon** [comment]
Superceded by https://github.com/micropython/micropython/pull/2189 . Further steps from there should be clean, one-change-at-time patches.
